File: services/typescript/express/src/routes/handlers/put.ts
import type { Request, RequestHandler, Response } from "express";
import { getLogger } from "telemetry";
import { AppErrors, sendError } from "../errors.js";
import { parseId, parseTask, type State } from "app";

export function putHandler(state: State): RequestHandler {
  return (req: Request, res: Response) => {
    const logger = getLogger();

    const id = parseId(req.params.id);
    if (!id.success) {
      logger.warn({ method: req.method, path: req.path }, "Invalid path");
      return sendError(res, AppErrors.InvalidPath);
    }

    const task = parseTask(req.body);
    if (!task.success) {
      logger.warn({ method: req.method, path: req.path }, "Invalid body JSON");
      return sendError(res, AppErrors.InvalidJsonBody);
    }

    const previous_task = state.tasks.get(id.data);

    if (previous_task === undefined) {
      logger.warn(
        { id: id.data, method: "PUT", path: req.path },
        "Task not found"
      );
      return sendError(res, AppErrors.TaskNotFound);
    }

    state.tasks.set(id.data, task.data);

    logger.info(
      {
        id: id.data,
        from_secret: previous_task.secret.length,
        to_secret: task.data.secret.length,
        from_operation: previous_task.operation.toString().toLowerCase(),
        to_operation: task.data.operation.toString().toLowerCase(),
        method: "PUT"
      },
      "Overwrote task"
    );

    return res.status(200).json(task.data);
  };
}

File: services/python/sanic/src/routes/handlers/put.py
import structlog
from uuid import UUID
from sanic import Request, json
from sanic.response import JSONResponse
from pydantic import ValidationError

from app.params import parse_task
from app.task import Task
from app.state import AppState
from routes.errors import send_error, AppErrors


async def put_handler(request: Request, id: UUID) -> JSONResponse:
    logger = structlog.get_logger()

    body = request.json
    if body is None:
        logger.warn("Invalid body JSON", method=request.method, path=request.path)
        return send_error(AppErrors.InvalidJsonBody)

    try:
        new_task: Task = parse_task(body)
    except ValidationError:
        logger.warn("Invalid body JSON", method=request.method, path=request.path)
        return send_error(AppErrors.InvalidJsonBody)

    state: AppState = request.app.ctx.state

    async with request.app.ctx.tasks_lock:
        task: Task | None = state.tasks.get(id)

        if task is None:
            logger.warn(
                "Task not found", id=str(id), method=request.method, path=request.path
            )
            return send_error(AppErrors.TaskNotFound)

        state.tasks[id] = new_task

    logger.info(
        "Overwrote task",
        id=str(id),
        from_secret=len(task.secret),
        to_secret=len(new_task.secret),
        from_operation=task.operation.lower(),
        to_operation=new_task.operation.lower(),
        method=request.method,
    )

    return json(new_task.model_dump())
